pytensor/graph/utils.py

```diff
@@ -420,3 +431,111 @@ def toposort(prereqs_d):
             "some orderings contain invalid elements."
         )
     return seq
+
+
+def _graph_replace(
+    outputs: Sequence["Variable"], replace: Dict["Variable", "Variable"]
+) -> Tuple[List["Variable"], Dict["Variable", "Variable"]]:
+    """Replace variables in ``outputs`` by ``replace`` in a single pass.
+
+    Parameters
+    ----------
+    outputs: Sequence[Variable]
+        Output graph
+    replace: Dict[Variable, Variable]
+        Replace mapping
+
+    Returns
+    -------
+    List["Variable"]
+        Output graph with subgraphs replaced
+    Dict["Variable", "Variable"]
+        Replacements that were not applied
+    """
+    from pytensor.graph.basic import Constant, condition_subset
+    from pytensor.graph.fg import FunctionGraph
+
+    # collect minimum graph inputs which is required to compute outputs
+    # and depend on replacements
+    # additionally remove constants, they do not matter in clone get equiv
+    conditions = [
+        c for c in condition_subset(outputs, replace) if not isinstance(c, Constant)
+    ]
+    # for the function graph we need the clean graph where
+    # inputs do not have owners
+    # this is exactly the reason to clone conditions
+    equiv = {c: c.clone(name=f"i-{i}") for i, c in enumerate(conditions)}
+    # some replace keys may dissapear
+    # the reason is they are outside the graph
+    # clone the graph but preserve the equiv mapping
+    fg = FunctionGraph(
+        conditions,
+        outputs,
+        # clone_get_equiv kwargs
+        copy_orphans=False,
+        copy_inputs=False,
+        memo=equiv,
+    )
+    # replace the conditions back
+    fg_replace = {equiv[c]: c for c in conditions}
+    # add the replacements on top of input mappings
+    fg_replace.update({equiv[r]: v for r, v in replace.items() if r in equiv})
+    # replacements have to be done in reverse topological order so that nested
+    # expressions get recursively replaced correctly
+
+    # some replacements may be initially outside the graph
+    # but later introduced by a replacement
+    # So far FunctionGraph does these replacements inplace it is thus unsafe
+    # apply them using fg.replace, it may change the original graph
+    non_fg_replace = {r: v for r, v in replace.items() if r not in equiv}
```

Comment:
Do we need to support this? Seems like we can leave to the user to call the original function twice. Why would he specify a replacement that replaces another replacement (IIRC)?

Reply:
Do you think it is too much magic? I would introduce a strict flag then to check if all replacements are used. Even in the current version

Reply:
Yeah I feel it's too much magic. You can just raise perhaps?